pytensor/tensor/random/basic.py

```diff
@@ -2057,113 +2059,182 @@ def make_node(self, rng, *args, **kwargs):
+            if p is None:
+                size = a.shape[:a_batch_ndim]
+            else:
+                size = np.broadcast_shapes(
+                    a.shape[:a_batch_ndim],
+                    p.shape[:p_batch_ndim],
+                )
 
-        Parameters
-        ----------
-        a
-            The array from which to randomly sample an element. If an int,
-            a sample is generated from `pytensor.tensor.arange(a)`.
-        size
-            Sample shape. If the given size is `(m, n, k)`, then `m * n *
-            k` independent samples are returned. Default is `None`, in
-            which case a single sample is returned.
-        replace
-            When `True`, sampling is performed with replacement.
-        p
-            The probabilities associated with each entry in `a`. If not
-            given, all elements have equal probability.
-        """
-        a = as_tensor_variable(a)
+        a = np.broadcast_to(a, size + a.shape[a_batch_ndim:])
+        if p is not None:
+            p = np.broadcast_to(p, size + p.shape[p_batch_ndim:])
 
-        if a.ndim == 0:
-            a = pytensor.tensor.arange(a)
+        a_indexed_shape = a.shape[len(size) + 1 :]
+        out = np.empty(size + core_shape + a_indexed_shape, dtype=a.dtype)
+        for idx in np.ndindex(size):
+            out[idx] = rng.choice(
+                a[idx], p=None if p is None else p[idx], size=core_shape, replace=False
+            )
+        return out
 
-        if p is None:
-            p = pytensor.tensor.type_other.NoneConst
 
-        if isinstance(replace, bool):
-            replace = pytensor.tensor.constant(np.array(replace))
+def choice(a, size=None, replace=True, p=None, rng=None):
+    r"""Generate a random sample from an array.
+
+
+    Parameters
+    ----------
+    a
+        The array from which to randomly sample an element. If an int,
+        a sample is generated from `pytensor.tensor.arange(a)`.
+    p
+        The probabilities associated with each entry in `a`. If not
+        given, all elements have equal probability.
+    replace
+        When `True`, sampling is performed with replacement.
+    size
+        Sample shape. If the given size is `(m, n, k)`, then `m * n *
+        k` independent samples are returned. Default is `None`, in
+        which case a single sample is returned.
+    """
+    a = as_tensor_variable(a)
+    a_size = a if (a.type.ndim == 0) else a.shape[0]
 
-        return super().__call__(a, p, replace, size=size, dtype=a.dtype, **kwargs)
+    if p is not None:
+        p = specify_shape(p, (a_size,))
 
+    if replace or size is None:
+        # In this case we build an expression out of simpler RVs
+        # This is equivalent to the numpy implementation:
+        # https://github.com/numpy/numpy/blob/2a9b9134270371b43223fc848b753fceab96b4a5/numpy/random/_generator.pyx#L905-L914
+        if p is None:
+            if rng is not None and isinstance(rng.type, RandomStateType):
```

Comment:
I don't get this distinction. What does this check do?

Reply:
integersRV works for Generator types and randint for the RandomState (and not vice versa). It will go away when we remove RandomStates next (together with randint)